### PR TITLE
886 add integration sync status progress

### LINKED ISSUE #886: Sync status integration progress
image

The last step here of `integration` could use a progress. The other steps say `n/total` records, integration surely could too! In my big data testing this is the slowest part... though perhaps I'm still getting loads of errors 😁.

### PR DESCRIPTION
**Pull request by fergie-nz**
Fixes #886 

# 👩🏻‍💻 What does this PR do? 
Adds logging and visual display for progress of integration step of syncing.

Note it only works on the initialising sync step - not syncing of already initialised setups.

This is because existing sync integration is run in a transaction.

# 🧪 How has/should this change been tested? 
You can test this by running branch [#886-Add](https://github.com/msupply-foundation/open-msupply/tree/886-testing-branch)
This adds an artificial delay to integration to simulate the visual logging progress in the front end.

## 💌 Any notes for the reviewer?
The tests aren't working on this branch.
They also aren't working on develop for sync (for me)

I'm not sure if I am exacerbating things, or if sync testing requires specific setup.

### CODE REVIEW

File: server/service/src/sync/synchroniser.rs [hunk 1]
```diff
@@ -183,16 +183,16 @@ impl Synchroniser {
...
         logger.done_step(SyncStep::PullRemote)?;
 
         // INTEGRATE RECORDS
         logger.start_step(SyncStep::Integrate)?;
-        //
-        let (upserts, deletes) =
-            integrate_and_translate_sync_buffer(&ctx.connection, is_initialised)
-                .map_err(SyncError::IntegrationError)?;
-        info!("Upsert Integration result: {:?}", upserts);
-        info!("Delete Integration result: {:?}", deletes);
```
Comment (fergie-nz) [inline server/service/src/sync/synchroniser.rs:195]: I assume these info flags are no longer necessary with logging added :)

**Review by mark-prins — CHANGES_REQUESTED** [review]
Looking great thanks!

much nicer having progress - thanks for wading through the back end code to sort this one out..

I have a fix for the test, and a suggestion.
Also had a look at the linked issue (#923) - that still crashes the server and I'd hoped that we could just show an integration error. Unfortunately `serde_json::from_str` does not return an error if one of the field translations panics, the server will simply die. I went down the rabbit hole for a bit and stopped.

Also wanted to have a look at getting the stats when running in operational mode, but have spent long enough!

**Comment by mark-prins** [comment]
if it's helpful - my changes are in [this branch](https://github.com/msupply-foundation/open-msupply/compare/886-testing-branch...886-testing-branch(integration))

**Review by mark-prins — APPROVED** [review]
👍  thanks for changes! looks great

**Review by andreievg — COMMENTED** [review]
Looks like i haven't quite submitted a review I did yesterday, a few comments. I think we should really just be doing progress for integration only and only for upserts

File: server/service/src/sync/synchroniser.rs [hunk 2]
```diff
@@ -235,20 +236,61 @@ pub fn integrate_and_translate_sync_buffer(
     > {
         let translators = all_translators();
         let table_order = pull_integration_order(&translators);
+        let step_progress = SyncStepProgress::Integrate;
 
         let sync_buffer = SyncBuffer::new(connection);
         let translation_and_integration = TranslationAndIntegration::new(connection, &sync_buffer);
         // Translate and integrate upserts (ordered by referential database constraints)
         let upsert_sync_buffer_records =
             sync_buffer.get_ordered_sync_buffer_records(SyncBufferAction::Upsert, &table_order)?;
```
[the remark this answers is not in the document]
Comment (andreievg) [reply server/service/src/sync/synchroniser.rs:247]: @fergie-nz 

As you've mentioned in telegram, this is actually not trivial:

https://github.com/msupply-foundation/open-msupply/compare/886-add-integration-sync-status-progress...886-add-integration-sync-status-progress(Example-of-option-mut)

File: server/service/src/sync/translation_and_integration.rs [hunk 3]
```diff
@@ -160,7 +140,35 @@ impl<'a> TranslationAndIntegration<'a> {
                     );
                 }
             }
+
+            if true == initialisation {
```
Comment (Chris-Petty) [inline server/service/src/sync/translation_and_integration.rs:144]: Nit: initialisation is a boolean, ya don't need to compare it!

Nit within a nit: https://en.wikipedia.org/wiki/Yoda_conditions. That said there isn't really a disadvantage it just reads a bit funny in English, and arguably there _are_ advantages as stated on wikipedia! I didn't know WordPress actually states it as a convention!

**Review by andreievg — CHANGES_REQUESTED** [review]
Just wanted to flag a point from my last comment as change request rather then nit. Thanks @Chris-Petty for mentioning that it's unclear if it's just a comment.

Basically as per the telegram chat, I would be happy to approve if this is implemented:
https://github.com/msupply-foundation/open-msupply/compare/886-add-integration-sync-status-progress...886-add-integration-sync-status-progress(Example-of-option-mut)(Simplified)

In summary:
* Only record integration progress for upserts during initialisation (don't record it during normal operations or for delete)
* Only show integration progress during initialisation (in UI), I haven't actually checked this bit

**Review by andreievg — APPROVED** [review]
Looks good, thanks @fergie-nz
